Function metadata

rongrong opened the pull request:
Introduce `FunctionMetadata` and use `FunctionMetadata` during planning. `FunctionMetadata` includes:
* Name
* ArgumentTypes
* ReturnType
* FunctionKind
* Deterministic
* CalledOnNullInput (call convention)

rongrong commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionMetadata.java:
```diff
@@ -0,0 +1,61 @@
+
+public class FunctionMetadata
+{
+    private final Signature signature;
```
`variableArity` is never used in planning. `getFunctionKind` is added as a function. I thought about whether to store a `Signature` or store `name`, `functionKind`, `argumentTyes` and `returnType` separately and I don't have a strong opinion one way or the other right now. Using `signature` here purely because it's less words to type. We can change if there's evidence that not storing signature is the better option.

highker commented on presto-main/src/main/java/com/facebook/presto/metadata/FunctionMetadata.java:
```diff
@@ -0,0 +1,61 @@
+
+public class FunctionMetadata
+{
+    private final Signature signature;
```
Why not using `FunctionHandle`?

rongrong replied:
You use a `FunctionHandle` to fetch `FunctionMetadata` `functionManager.getFunctionMetadata(FunctionHandle)`. It would be really strange if metadata is a wrapper on handle... What's the point... The implementation of this will likely change in the future, as I mentioned in https://github.com/prestodb/presto/pull/12570#discussion_r271081765. But it's not likely to ever hold a `FunctionHandle` as a member. 😛

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java:
```diff
@@ -266,23 +268,26 @@ private TopNRowNumberNode convertToTopNRowNumber(WindowNode windowNode, int limi
-        private static boolean isRowNumberSignature(Signature signature)
+        private static boolean isRowNumberMetadata(FunctionMetadata functionMetadata)
         {
-            return signature.equals(ROW_NUMBER_SIGNATURE);
```
Do we wanna introduce some helper functions to simplify these comparisons...

highker commented on presto-main/src/main/java/com/facebook/presto/sql/planner/ExpressionInterpreter.java:
```diff
@@ -899,10 +897,9 @@ protected Object visitFunctionCall(FunctionCall node, Object context)
             for (int i = 0; i < argumentValues.size(); i++) {
                 Object value = argumentValues.get(i);
-                if (value == null && function.getArgumentProperty(i).getNullConvention() == RETURN_NULL_ON_NULL) {
+                if (value == null && !functionMetadata.isCalledOnNullInput()) {
```
In general we may prevent users from using `getNullConvention`. Otherwise, #12483 can still rely on this function call without failing

rongrong replied:
Can you elaborate? I don't understand what you mean.

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/WindowFilterPushDown.java:
```diff
@@ -266,23 +268,26 @@ private TopNRowNumberNode convertToTopNRowNumber(WindowNode windowNode, int limi
+        private static boolean isRowNumberMetadata(FunctionMetadata functionMetadata)
         {
-            return signature.equals(ROW_NUMBER_SIGNATURE);
+            return functionMetadata.getName().equals(ROW_NUMBER_SIGNATURE.getName())
```
Can we have `ROW_NUMBER_FUNCTION_METADATA` and directly do `functionMetadata.equals(ROW_NUMBER_FUNCTION_METADATA)` ?

rongrong replied:
The current situation is definitely not ideal. But clean up all signature related issues is an ongoing process (refer to be previous dozen commits). We'll eventually get them all.

wenleix commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ParametricScalarImplementation.java:
```diff
@@ -629,6 +629,17 @@ else if (Stream.of(annotations).anyMatch(BlockPosition.class::isInstance)) {
+                    .map(ArgumentProperty::getNullConvention)
+                    .filter(nullConvention -> nullConvention != RETURN_NULL_ON_NULL)
+                    .count();
+            checkCondition((header.getHeader().isCalledOnNullInput() && valueArgumentsReturnNullOnNullCount > 0)
```
I personally prefer to only use `checkCondition` or `checkState` when the condition is simple. In this case, I would just explicitly write if and throw the exception :)

rongrong replied:
Too much typing to throw an exception 😛

wenleix commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/annotations/ParametricScalarImplementation.java:
```diff
@@ -629,6 +629,17 @@ else if (Stream.of(annotations).anyMatch(BlockPosition.class::isInstance)) {
                     }
                 }
             }
+            long valueArgumentsReturnNullOnNullCount = argumentProperties.stream()
```
nit: Is it actually  counting *not* return null on null ? :) -- seems not matching the variable name ?  :)

wenleix commented on presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayDistinctFromOperator.java:
```diff
@@ -32,7 +32,7 @@
 import static com.facebook.presto.util.Failures.internalError;
 
-@ScalarOperator(IS_DISTINCT_FROM)
+@ScalarOperator(value = IS_DISTINCT_FROM)
```
nit: unrelated change ? :)

wenleix commented on presto-main/src/main/java/com/facebook/presto/sql/InterpretedFunctionInvoker.java:
```diff
@@ -70,7 +71,13 @@ private Object invoke(ScalarFunctionImplementation function, ConnectorSession se
             ArgumentProperty argumentProperty = function.getArgumentProperty(i);
             if (argumentProperty.getArgumentType() == VALUE_TYPE) {
-                if (function.getArgumentProperty(i).getNullConvention() == USE_NULL_FLAG) {
+                if (function.getArgumentProperty(i).getNullConvention() == RETURN_NULL_ON_NULL) {
```
hidden bug? :)

rongrong replied:
This won't happen before because in `ExpressionIntepreter` it's checking whether each argument's call convention is `RETURN_NULL_ON_NULL` and if that's the case & the value is null it would optimize it there so the function will never be invoked. Now that we change this call convention in metadata to be per function rather than per argument, some of those won't be optimized out and the actual execution need to take care of doing the right thing.

rongrong commented:
> Still, might want to explain a bit we want to make `deterministic` a function level property (so it's in `FunctionMetadata`, instead of each implementation's property (so moved out from `ScalarFunctionImplementation`)?

Whether a function is deterministic is part of SQL spec so it maps to a SqlFunction conceptually. Though I agree that with generic types, a SqlFunction is not necessarily one function, and theoretically you could have different implementation for different input types with different configuration of "deterministic", but that's super confusing. They'd better be different functions that that point.

rongrong commented:
Rebased and ready for review.

rongrong commented:
I replaced the signature with the parts.

rongrong commented:
> Approving for commits other than adding "display name" for `CallExpression`. But given `CallExpression` is in SPI, I don't have strong opinion towards whether having "display name" as another field or just demangling the signature name for display name. If it is not in SPI, I would definite vote for demangling. Are we going to use "display name" for logic use like comparison? Or simply for display?
> 
> May have @wenleix to take another pass.

@dain  added a "nameHint" to `CallExpression` in his original PR. I thought that was confusing and not necessary so I didn't take that. This is not for mangling / demangling operators (I don't care whether we print the mangled version or unmangled version). It's because I don't want to have a call to `FunctionManager` to fetch `FunctionMetadata` just to get the name of the call in `toString`.

rongrong commented:
> Commits up to "Add call convention to FunctionMetadata" look good to me. How did you figure out all the `calledOnNullInput = true` functions?

@highker  @wenleix  I added validation to make sure this setting is consistent with argument null conventions so I was relying on test failures to catch them. If we introduced functions without tests we might be in trouble... Do you guys have better suggestions?

wenleix commented:
@rongrong : One thought is to have a tests that iterate over all the functions and check if the setting is consistent. Similar to `testOperatorsImplemented`: https://github.com/prestodb/presto/blob/b9b85be63011dc16e2a77279c35e07cf8cff4ce0/presto-main/src/test/java/com/facebook/presto/type/TestTypeRegistry.java#L266-L283

However I am not sure how feasible it is to apply to this case. If not, feel free to ignore this comment :)

rongrong commented:
> @rongrong : One thought is to have a tests that iterate over all the functions and check if the setting is consistent. Similar to `testOperatorsImplemented`:
Don't know how to do that since we don't really have an interface to let functions describe what type of inputs they accept. I'll just assume all functions are introduced with tests, which doesn't sound too unreasonable...
